### PR TITLE
enable the userData websocket endpoint to reconnect gracefully

### PR DESCRIPTION
**Pull request by gkaindl**
Thanks for your module!

I noticed a problem related to the userData websocket endpoint, which I think is probably the same as the ones mentioned in #123  and #110.

The userData endpoint requires to obtain, keep alive, and delete a listenKey via the REST API, but this requirement isn't taken into account with the auto-reconnecting websocket.

1) The keep-alives, which are sent from an interval, can very well fail, especially with a spotty internet connection -> that would then cause an unhandled promise rejection, crashing the entire app from the internals of binance-api-node module.

2) I was frequently getting the same "This listenKey does not exist." error from the binance API as the described in the aforementioned issues. I think it's related to either the keep-alive requests not going through for whatever reason (maybe timing out, etc...), or it's simply a problem on binance's end. In any case, this would also crash the app then, with no good way to handle the error, since the module currently assumes the listenKey doesn't need to be changed after reconnect.

In this pull request, I'm trying to enable the expected auto-reconnecting behavior, even if the REST API requests fail, e.g. by hiding errors happening within the internals, but still exposing errors happening during the endpoint setup/tear-down via the client app api.

### CODE REVIEW

**Comment by balthazar** [comment]
Hey, thanks!

So if I understand correctly, you think the issues com from the fact the keep-alive requests are failing and thus the listenKey is inexistant on Binance end? Wouldn't refactoring the `keepStreamAlive` method to recall itself in case of error and catching potential failures of the `closeDataStream` be enough rather than doing your implementation?

**Comment by balthazar** [comment]
I think #117 is most likely what the issue was actually

**Comment by gkaindl** [comment]
yes, that, and that the call to keep-alive, which happens from the interval, could fail as well, such as if your internet connection drops out intermittently (bad wireless reception, etc...). Since the keep-alive REST request is happening periodically, I think its sensible to guard against such failures as well, since they'd happen entirely in the background from a client app's perspective, and there isn't a way to catch it.

So basically, with this patch, you can connect to the userData endpoint, then let the network drop out (put your computer to sleep, etc...), and it'll reconnect automatically when internet is available again, mimicking the behavior of the other websocket endpoints.

**Comment by balthazar** [comment]
Since the recommended keep-alive is 30 minutes and we do it every 30 seconds currently, it seems unlikely 60 consecutive calls would fail.

The sleep and wake use-case seems pretty limited to me as I don't see someone running a bot on their laptop and expecting it to work correctly after being put to sleep for a prolonged period of time.

Catching the failure of the close call seems like the only important step atm, unless a keepAlive failure is also subject to crashing the app and would need to be catched too

**Comment by gkaindl** [comment]
Well, at least a failure of the keepAlive call needs to be catched too, because it'd currently crash the app.

Fair enough about not wanting to support the use-case, though – I'm using binance-api-node not only for a bot that's run server-side, but also for a locally run electron app on a Mac, where letting apps resume gracefully after sleep/wake is generally expected.

Maybe another solution would be to provide some sort of an error callback or so, because in such circumstances, there wouldn't even be a way to detect that the userData endpoint did not resume automatically, once keepAlive failures don't crash the app anymore, so that case might go unnoticed by client app developers.

**Comment by balthazar** [comment]
Ok I see. Can you please take a look at https://github.com/HyperCubeProject/binance-api-node/commit/d41eaf94df2927baec61a19b3ac3acf06be5c2bf and check if it would at least fix potential crashes following errored requests?

So the only thing left would be the handling of prolonged disconnects somehow. How feasible would it be for you to hook into an event telling you the app has been foregrounded and recreate the socket?

**Comment by gkaindl** [comment]
Looks good to mitigate the uncatchable rejections on keepAlive or close errors!

Yes, reconnecting upon wake can be done, that's not a problem. The issue is more that the userData websocket behaves different from the others in that it does not automatically reconnect properly after such a disconnection. 

If it's going to be implemented that way, I'd suggest to add this caveat to the documentation of this endpoint, because there isn't really a way for the client app to notice the disconnection under these circumstances, and it's quite different from all the other websocket endpoints in this respect (the userData websocket is quite low volume, so doing a watchdog to reconnect if nothing has been received for x minutes isn't really viable).

**Comment by balthazar** [comment]
Doesn't Binance sends a specific error code to the keepAlive request in case the listenKey doesn't exist anymore that we could use to throw a specific error people would be able to catch on their side and recall the method? I don't think we can implement the auto-reconnect in the library as I'd like to keep the returned function to clear the refresh interval and close the stream

**Comment by numerized** [comment]
Hello guys, any news on this ?

**Comment by balthazar** [comment]
Have you even tried this branch to see if it resolved the disconnection issue?

**Comment by numerized** [comment]
I'm trying it right now yes. tomorrow I'll update you. Thanks for asking!

**Comment by numerized** [comment]
Might take more time to test it, Binance was under maintenance.
So I'll know more about this fix tomorrow then, I need 24hours to witness disconnect usually.

**Comment by numerized** [comment]
After the maintenance the socket reconnected gracefully at least!
I'll let this process run and let you know tomorrow if it's still up!
Thanks it's promising!

**Comment by numerized** [comment]
Confirmed. The fix is fully working!

**Comment by balthazar** [comment]
Stop spamming the issue please 🙂 

@gkaindl would it be easy for you to create a unit test for it by killing the socket and see if it reconnects? I'd probably like to refactor a bit the code in the future without risking to break functionality.

Also try to rebase from master

**Comment by gkaindl** [comment]
@balthazar I'll see if I can find the time to do it this weekend!

**Comment by numerized** [comment]
Hello @gkaindl any progress on this brother?

**Comment by numerized** [comment]
Hi Guys, your .MSG socket are still hanging up without any error message.
And now I had this issue with this branch on 8 processes.

What can we do?

**Comment by numerized** [comment]
Hello guys, can we finalize this pull request ?? I'm testing it ok right now. No more disconnections, my previous post might be a mistake on my test protocol

**Comment by balthazar** [comment]
Alright without talking about unit tests, maybe if @gkaindl fixes the merge conflicts I can get that landed

**Comment by numerized** [comment]
10 days uptime here and no problem on the msg socket.

**Comment by numerized** [comment]
@gkaindl @balthazar 
Can we merge finally ?
I'm not advanced enough to do the job and I really want this to be deployable easily without having to compile this branch on my deploy script.

Thanks guys.

**Comment by balthazar** [comment]
Implemented in `0.9.0`
